### PR TITLE
download password protected public shared files using pre-signed url

### PR DESCRIPTION
**Pull request by C0rby**
## Description
Replaced the blob download with normal downloads using pre-signed urls provided by the backend.

## Related Issue

- Fixes <issue_link>

## Motivation and Context

The blob download had some disadvantages like delayed download for big files and high memory consumption.

## How Has This Been Tested?

- locally using the OC10 backend

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: apps/files/src/store/state.js
Comment (LukasHirt) [inline apps/files/src/store/state.js:23]: Is the download URL available only for public links or also for private files?

Comment (kulmann) [reply apps/files/src/store/state.js:23]: Hehehehehe, I asked the same thing in the chat. It is not. I would love to have all download urls already by the server. Then we could get rid of the url signing in the sdk and would not need to make any difference between public and private context.

File: src/plugins/web.js
Comment (kulmann) [inline src/plugins/web.js:88]: I would like to get rid of this code duplication with the `a` element.

Comment (kulmann) [reply src/plugins/web.js:88]: Can be moved out into its own method.

File: apps/files/src/store/state.js
Comment (LukasHirt) [reply apps/files/src/store/state.js:23]: Hehe. Exactly my thoughts!

File: src/plugins/web.js
Comment (C0rby) [reply src/plugins/web.js:88]: Should I tackle this here in this PR?

Comment (kulmann) [reply src/plugins/web.js:88]: Yes please :-)

**Comment by C0rby** [comment]
The tests are failing because the backends which are used to test against do not contain the API changes yet.

**Comment by C0rby** [comment]
Here is a PR which fixes some of the test failures: https://github.com/owncloud/ocis/pull/1785

**Comment by C0rby** [comment]
I don't get it...these tests only fail in the CI. Locally they run successfully.
@kulmann @LukasHirt @phil-davis, any idea?

## webUIBasic-chrome
webUIPreview/imageMediaViewer.feature:15
webUIPreview/imageMediaViewer.feature:14
webUIPreview/imageMediaViewer.feature:65
webUIPreview/imageMediaViewer.feature:13
webUIPreview/imageMediaViewer.feature:145
webUIPreview/imageMediaViewer.feature:127
webUIPreview/imageMediaViewer.feature:54
webUIPreview/imageMediaViewer.feature:57
webUIPreview/imageMediaViewer.feature:74

## webUIOCISBasic-chrome
webUIPreview/imageMediaViewer.feature:14
webUIPreview/imageMediaViewer.feature:65
webUIPreview/imageMediaViewer.feature:13
webUIPreview/imageMediaViewer.feature:54
webUIPreview/imageMediaViewer.feature:57
webUIPreview/imageMediaViewer.feature:74

File: packages/web-app-files/src/mixins/actions/fetch.js
Comment (kulmann) [inline packages/web-app-files/src/mixins/actions/fetch.js:33]: Please use the server url from the store.
Add `import { mapGetters } from 'vuex'` as import and `...mapGetters(['configuration']),` within the computed properties. The you can use the server url like this: `this.configuration.server`. It is guaranteed to have a trailing slash.
The `$client` naturally has the same url, because it is initialized with it. But you should grab it from the source, not from an intermediate object.

Comment (kulmann) [reply packages/web-app-files/src/mixins/actions/fetch.js:33]: And yes, it is already super ugly to use the webdav url like this from the client. ;-) But that can be fixed later. For the server url it's a quick win.

File: packages/web-app-media-viewer/src/App.vue
Comment (kulmann) [inline packages/web-app-media-viewer/src/App.vue:170]: Could you move this entire block into the `$_loader_getDavFilePath` mixin and just provide the entire `activeMediaFile` instead of the path? Then you're also able to get rid of the path traversal with `../../../`. The URL format is well known to us, I'd prefer string replaces instead of traversals if possible.

File: packages/web-runtime/src/plugins/web.js
Comment (kulmann) [inline packages/web-runtime/src/plugins/web.js:36]: use the `configuration.server` url again instead.

Comment (kulmann) [reply packages/web-runtime/src/plugins/web.js:36]: And get rid of the commented line please :-)

**Review by kulmann — CHANGES_REQUESTED** [review]
Some small changes needed. Thanks for pushing this change! It's a really good step forward. 🚀

**Review by kulmann — CHANGES_REQUESTED** [review]
Found something ugly. The backend returns the following downloadURL:
`/px1/remote.php/dav/public-files/hXzyzP24D3YC2hQ/Screenshot 2021-03-16 at 13.35.31.png?signature=7e922b4728b71f969c0024d1e826141591615a2a7543f45a67daed2237b1b043&expires=2021-03-16T14%3A35%3A01%2B00%3A00`
and it gets concatenated with the server url, which is `http://host.docker.internal/px1/`, resulting in a duplicated `px1` path. My ownCloud is in a subfolder, which is a valid setup.
Does this require a core change again? 👀

**Comment by C0rby** [comment]
Ah damn it.. Possible... let me check

**Comment by C0rby** [comment]
@kulmann, what do you think would make more sense?

Put the full url into `downloadURL` like: `http://localhost:8080/owncloud/remote.php/dav/public-files/hXzyzP24D3YC2hQ/Screenshot 2021-03-16 at 13.35.31.png?signature=7e922b4728b71f969c0024d1e826141591615a2a7543f45a67daed2237b1b043&expires=2021-03-16T14%3A35%3A01%2B00%3A00`
Or always only put `/remote.php/dav/public-files/hXzyzP24D3YC2hQ/Screenshot 2021-03-16 at 13.35.31.png?signature=7e922b4728b71f969c0024d1e826141591615a2a7543f45a67daed2237b1b043&expires=2021-03-16T14%3A35%3A01%2B00%3A00`

**Comment by kulmann** [comment]
> @kulmann, what do you think would make more sense?
> 
> Put the full url into `downloadURL` like: `http://localhost:8080/owncloud/remote.php/dav/public-files/hXzyzP24D3YC2hQ/Screenshot 2021-03-16 at 13.35.31.png?signature=7e922b4728b71f969c0024d1e826141591615a2a7543f45a67daed2237b1b043&expires=2021-03-16T14%3A35%3A01%2B00%3A00`
> Or always only put `/remote.php/dav/public-files/hXzyzP24D3YC2hQ/Screenshot 2021-03-16 at 13.35.31.png?signature=7e922b4728b71f969c0024d1e826141591615a2a7543f45a67daed2237b1b043&expires=2021-03-16T14%3A35%3A01%2B00%3A00`

Would be super handy to the full download url already available. But is that again more error prone (regarding server side configuration)? If yes, that it's not an issue having to prefix the path with the server url from the web config.

**Comment by C0rby** [comment]
Yeah, I think just sending the relative URL is more flexible for more deployment scenarios.

**Comment by jnweiger** [comment]
> Yeah, I think just sending the relative URL is more flexible for different deployment scenarios.

Why do we go for flexibility? I'd argue that a full URL is more robust, as no errors can be made during URL-assembly.

**Review by kulmann — APPROVED** [review]
LGTM now 🚀

**Comment by kulmann** [comment]
CI is failing for `webUIPreview/imageMediaViewer.feature:24` and `webUIPreview/imageMediaViewer.feature:33` because the core changes are not in master, yet. You'd need to switch to the release branch and respective commit id in the `.drone.env` file for now. I don't know if this is fine for @phil-davis and @individual-it though.

**Comment by individual-it** [comment]
you can change the branch in `.drone.env` to prove that the change is working, but I would not merge it that way, better to get the change into core first or add the failing test to the expected failures

**Comment by micbar** [comment]
@individual-it we are a bit in a cross-blocking situation.

The change in core is committed to the `release-10.7.0` branch.
It will be merged back to master after the release has been done.

Currently the server release is waiting for a new web release with this fix here.

We need to unblock here.

Raising to P1

**Pull request merged**
